### PR TITLE
chore(aws_cloudwatch_logs sink): re-write aws cloudwatch logs sink in new style

### PR DESCRIPTION
**Pull request by fuchsnj**
closes https://github.com/vectordotdev/vector/issues/9452

### CODE REVIEW

File: src/sinks/aws_cloudwatch_logs/service.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+            create_missing_group,
+            create_missing_stream,
+            token: None,
+            token_rx: None,
+        }
+    }
+
+    pub fn process_events(&self, events: Vec<InputLogEvent>) -> Vec<Vec<InputLogEvent>> {
+        let now = Utc::now();
+        // Acceptable range of Event timestamps.
```
Comment (tobz) [inline src/sinks/aws_cloudwatch_logs/service.rs:208]: Is there a reason we can't do this at a higher level? Seems like it'd be more efficient to never hold on to these events in the first place if the filter criteria are hard-coded/known in advance.

File: src/sinks/aws_cloudwatch_logs/service.rs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+
+        // Sort by timestamp
+        events.sort_by_key(|e| e.timestamp);
+
+        info!(message = "Sending events.", events = %events.len());
+
+        let mut event_batches = Vec::new();
+        if events.is_empty() {
+            // This should happen rarely.
+            event_batches.push(Vec::new());
```
Comment (tobz) [inline src/sinks/aws_cloudwatch_logs/service.rs:225]: Do we _have _ to send an empty batch? 🤔

File: src/sinks/aws_cloudwatch_logs/service.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+                    self.token = token;
+                    self.token_rx = None;
+                }
+                Err(_) => {
+                    // This case only happens when the `tx` end gets dropped due to an error
+                    // in this case we just reset the token and try again.
+                    self.token = None;
+                    self.token_rx = None;
+                }
+            }
```
Comment (tobz) [inline src/sinks/aws_cloudwatch_logs/service.rs:279]: I think you can collapse this down to something like:

```rust
self.token = ready!(rx.poll_unpin(cx)).ok();
self.token_rx = None;
```

File: src/sinks/aws_cloudwatch_logs/sink.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        let service = self.service;
+        let acker = self.acker;
+
+        input
+            .filter_map(|event| {
+                future::ready(match request_builder.build(event) {
+                    Ok(maybe_req) => maybe_req.map(Ok),
+                    Err(err) => Some(Err(err)),
+                })
+            })
```
Comment (tobz) [inline src/sinks/aws_cloudwatch_logs/sink.rs:44]: This could be:

```rust
.filter_map(|event| async { request_builder.build(event).transpose() })
```

**Review by tobz — COMMENTED** [review]
A few general comments/questions, but otherwise looking pretty good. 👍🏻

**Review by tobz — APPROVED** [review]
Subsequently approving this per a side discussion about my comments really being nits, not blocking items.

File: src/sinks/aws_cloudwatch_logs/service.rs (hunk 1 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/service.rs:208]: No reason, I'll move this up so they are filtered out immediately.

File: src/sinks/aws_cloudwatch_logs/service.rs (hunk 2 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/service.rs:225]: I don't understand why we originally had this, but looking through the code I don't see a reason to keep it. I'm moving the date filtering logic to a higher level, which will make this branch impossible to reach anyway, so I'll just delete it.

File: src/sinks/aws_cloudwatch_logs/service.rs (hunk 1 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/service.rs]: fixed in ad628ee0140d02cb601f09e1145031ea7bd15895

File: src/sinks/aws_cloudwatch_logs/service.rs (hunk 2 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/service.rs]: fixed in ad628ee0140d02cb601f09e1145031ea7bd15895

File: src/sinks/aws_cloudwatch_logs/service.rs (hunk 3 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/service.rs]: fixed in ad628ee0140d02cb601f09e1145031ea7bd15895

File: src/sinks/aws_cloudwatch_logs/sink.rs (hunk 4 above)
Comment (fuchsnj) [reply src/sinks/aws_cloudwatch_logs/sink.rs]: fixed in ad628ee0140d02cb601f09e1145031ea7bd15895

**Pull request merged**